### ISSUE TITLE
support gpu acceleration on multiple platforms

### ISSUE THREAD

**Issue by cameronraysmith** [pinellolab/pyrovelocity#6]
To support GPU utilization on Apple M1 or M2 ( https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/ ) pyrovelocity would need to support pytorch `>=1.12`​. For reference the latest stable version is ​`1.12.1​`.

The library currently supports ​`1.8.*​`. 

With, for example, the following channels ([note ​`nodefaults​`](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html?highlight=nodefaults#creating-an-environment-file-manually), which helps to avoid confusion of dependency versions between conda-forge and anaconda package repositories)

```yml
channels:
- conda-forge
- bioconda
- nodefaults
```

and inclusion of dependency on

```yml
  - pytorch=1.8.*
```

in the default environment configuration file, then when `  - pytorch>=​1.12` (at upgrade from current dependencies across the `1.12` threshold), GPU acceleration should work by default on the M1/2.

Relative to the above dependency in the package `environment.yml`, 

<details>
<summary>environment.yml</summary>

```yaml
name: pyrovelocity

channels:
  - conda-forge
  - bioconda
  # https://stackoverflow.com/a/71110028/446907
  # We want to have a reproducible setup, so we don't want default channels,
  # which may be different for different users. All required channels should
  # be listed explicitly here.
  - nodefaults

dependencies:
  - python=3.8
  - mamba=0.27.0
  - conda-lock=1.1.3
  - leidenalg=0.9.0
  - pyro-ppl=1.6.0
  - pip=22.2.2
  - seaborn=0.11.2
  - scvelo=0.2.4
  - scvi-tools=0.13.0
  - pytorch-lightning=1.3.0
#  - pytorch-gpu=1.8.*
#  - pytorch=1.8.*=*cuda*
  - pytorch=1.8.*
  - scikit-misc=0.1.4
  - torchmetrics=0.5.1
  - h5py=3.7.0
  - anndata=0.7.5
  - adjusttext=0.7.3
  - astropy=5.1
  - pip:
    - cospar==0.2.1
```

</details>

including a note such as the following

```markdown
To enable GPU acceleration on linux platforms, you may attempt to execute a command similar to

conda install -y pytorch-gpu=1.8.* -c conda-forge

if your environment is compatible, for example, with 
[conda virtual package](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html) 
`__cuda >= 11`. 

Note that the `pytorch-gpu` package pointer only works on linux platforms. 
An alternative is to specify constraints such as `pytorch=1.8.*=*cuda*`. 
Please see information regarding associated dependencies in

- [conda-forge/pytorch-gpu:1.8.0](https://anaconda.org/conda-forge/pytorch-gpu/files?version=1.8.0) 
- [conda-forge/cudatoolkit:11.2.2](https://anaconda.org/conda-forge/cudatoolkit/files?version=11.2.2)
- [nvidia cuda toolkit <--> driver version compatibility](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-major-component-versions__table-cuda-toolkit-driver-versions)
```

should allow anyone to enable GPU acceleration on linux platforms. See references to one dependency chain from [pytorch-gpu](https://anaconda.org/conda-forge/pytorch-gpu/files?version=1.8.0) to [cudatoolkit](https://anaconda.org/conda-forge/cudatoolkit/files?version=11.2.2) below.

**Comment by cameronraysmith** [comment]
Note that
https://anaconda.org/conda-forge/pytorch-gpu/files?version=1.8.0 and environment with `cuda 11.2` and `python 3.8` would attempt to install
https://anaconda.org/conda-forge/pytorch-gpu/1.8.0/download/linux-64/pytorch-gpu-1.8.0-cuda112py38h0bbbad9_1.tar.bz2

<h4 class="modal-title" id="myModalLabel" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-style: normal; font-weight: 400; line-height: 1.4; text-rendering: optimizelegibility; font-size: 1.4375rem; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">linux-64/pytorch-gpu-1.8.0-cuda112py38h0bbbad9_1.tar.bz2</h4><p class="subheader" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-weight: 400; text-rendering: optimizelegibility; line-height: 1.4; color: rgb(111, 111, 111); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">No Description</p><div style="box-sizing: border-box; margin: 0px; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Uploaded | Mon Apr 12 13:01:42 2021
-- | --
md5 checksum | 5744fcaa29a36524a7268fc8c33e3cd6
arch | x86_64
build | cuda112py38h0bbbad9_1
build_number | 1
depends | pytorch 1.8.0 cuda112py38h3bc52bc_1
license | BSD-3-Clause
license_family | BSD
machine | x86_64
operatingsystem | linux
platform | linux
subdir | linux-64
target-triplet | x86_64-any-linux
timestamp | 1618214098213

</div>

This is just a reference to `pytorch 1.8.0 cuda112py38h3bc52bc_1` in https://anaconda.org/conda-forge/pytorch/files?version=1.8.0 at
https://anaconda.org/conda-forge/pytorch/1.8.0/download/linux-64/pytorch-1.8.0-cuda112py38h3bc52bc_1.tar.bz2
<h4 class="modal-title" id="myModalLabel" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-style: normal; font-weight: 400; line-height: 1.4; text-rendering: optimizelegibility; font-size: 1.4375rem; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">linux-64/pytorch-1.8.0-cuda112py38h3bc52bc_1.tar.bz2</h4><p class="subheader" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-weight: 400; text-rendering: optimizelegibility; line-height: 1.4; color: rgb(111, 111, 111); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">No Description</p><div style="box-sizing: border-box; margin: 0px; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Uploaded | Mon Apr 12 13:01:30 2021
-- | --
md5 checksum | a18469847f5607699658769bf0eca4b2
arch | x86_64
build | cuda112py38h3bc52bc_1
build_number | 1
depends | __glibc >=2.17, __glibc >=2.17,<3.0.a0, _openmp_mutex >=4.5, cffi, cudatoolkit 11.2\|11.2.*, cudnn >=8.1.0.77,<9.0a0, future, libblas * *_mkl, libblas >=3.9.0,<4.0a0, libgcc-ng >=7.5.0, libprotobuf >=3.15.8,<3.16.0a0, libstdcxx-ng >=7.5.0, magma >=2.5.4,<2.5.5.0a0, mkl >=2020.4,<2021.0a0, nccl >=2.8.4.1,<3.0a0, ninja, numpy >=1.17.5,<2.0a0, python >=3.8,<3.9.0a0, python_abi 3.8.* *_cp38, sleef >=3.5.1,<4.0a0, typing_extensions
has_prefix | True
license | BSD-3-Clause
license_family | BSD
machine | x86_64
operatingsystem | linux
platform | linux
subdir | linux-64
target-triplet | x86_64-any-linux
timestamp | 1618213801933

</div>

which depends on `cudatoolkit 11.2|11.2.*`​. This may resolve to 
https://anaconda.org/conda-forge/cudatoolkit/files?version=11.2.2

<h4 class="modal-title" id="myModalLabel" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-style: normal; font-weight: 400; line-height: 1.4; text-rendering: optimizelegibility; font-size: 1.4375rem; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">linux-64/cudatoolkit-11.2.2-hbe64b41_10.tar.bz2</h4><p class="subheader" style="box-sizing: border-box; margin: 0.2rem 0px 0.5rem; padding: 0px; font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-weight: 400; text-rendering: optimizelegibility; line-height: 1.4; color: rgb(111, 111, 111); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">No Description</p><div style="box-sizing: border-box; margin: 0px; padding: 0px; color: rgb(34, 34, 34); font-family: &quot;Museo Sans Rounded&quot;, Lato, &quot;Helvetica Neue&quot;, Helvetica, Roboto, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Uploaded | Thu May 19 21:00:44 2022
-- | --
md5 checksum | 363bb01846b2a4b0a575f4793cf78d27
arch | x86_64
build | hbe64b41_10
build_number | 10
constrains | __cuda >=11
depends | __glibc >=2.17,<3.0.a0, libgcc-ng >=10.3.0, libstdcxx-ng >=10.3.0
license | LicenseRef-NVIDIA-End-User-License-Agreement
machine | x86_64
operatingsystem | linux
platform | linux
subdir | linux-64
target-triplet | x86_64-any-linux
timestamp | 1652993270938

</div>

which depends upon detection of [conda virtual package](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html) `__cuda >=11`.

**Comment by cameronraysmith** [comment]
This issue should be closed when pytorch is upgraded to 1.12 or greater.

**Comment by qinqian** [comment]
We constrain pytorch=1.8.* because the latest pyro 1.8.1+06911dc with pytorch 1.12.1 will not support model-side sequential enumeration as in [pyro code](https://github.com/pyro-ppl/pyro/blob/85a288ab7dd352d288352b4456576903c4b89409/pyro/util.py#L300) and throw the error below: 

NotImplementedError: At site 'cell_gene_state', model-side sequential enumeration is not implemented. Try parallel enumeration or guide-side enumeration.

As above, this issue will be closed when pyro supports model-side sequential enumeration.

> This issue should be closed when pytorch is upgraded to 1.12 or greater.

**Comment by cameronraysmith** [comment]
This issue will be addressed when all dependencies are able to be upgraded to latest stable versions.
